BOLT 12: remove `send_invoice` offer, use a straight `invoice_request`.

rustyrussell opened the pull request:
Instead of a kind of dummy offer which asks you to send an invoice,
use a dummy invoice request, since it now contains all the offer
fields anyway.  It simplifies things a little, but there's the
following fallout:

1. We use the presence of `offer_node_id` to distinguish
   invoice-request-for-offer and invoice-request-no-offer, which makes
   sense, since the vendor doesn't know that.
2. We rename `invoice_request_payer_key` to `invoice_request_payer_id`.  In
   the case of a public node making a non-offer invoice_request, it's
   their id (as they are the payer), but it can omitted using blinded
   paths (like the offer_node_id in offer).
3. The invoice_request has no signature if it's not an offer response.
   This is makes them as small like the old send_invoice offer.
4. We don't allow invoice_request in other currencies (could in future):
   must specify in msats (this issue is expanded in the error section).
5. Actually tell them they're supposed to request an invoice when they
   receive an `invoice_request` for an offer!  (They don't have to if it's
   a non-offer one, they would be expected to check with user).
6. No signatures on invoices if they're non-offer, but allow a verification
   code.  If the user presents an offer to the ATM in the first place, this
   can be skipped, but requires it have a camera, and more interaction.

Here are examples of the offer-request-invoice flow, and the
request-invoice flow:

### offer-request-invoice: Rusty buys a coffee from Starblocks

1. Scans lno...

   * offer_description: 1 cup coffee.
   * offer_amount: 1000000msat
   * offer_issuer: Starblocks starblocks.acinq.co
   * offer_node_id: starblocks-nodeid

2. Rusty sends onionmessage to starblocks-nodeid with reply_path back Rusty, and
invoice_request:

   * invoice_request_payer_info: 23456...
   * offer_description: 1 cup coffee.
   * offer_amount: 1000000msat
   * offer_issuer: Starblocks starblocks.acinq.co
   * offer_node_id: starblocks-nodeid
   * invoice_request_payer_id: rusty-tmpkey
   * invoice_request_payer_note: love that coffee!
   * signature: SIG(^^, rusty-tmpkey)
   
3. Starblocks replies with invoice:

   * invoice_request_payer_info: 23456...
   * offer_description: 1 cup coffee.
   * offer_amount: 1000000msat
   * offer_issuer: Starblocks starblocks.acinq.co
   * offer_node_id: starblocks-nodeid
   * invoice_request_payer_id: rusty-tmpkey
   * invoice_request_payer_note: love that coffee!
   * invoice_paths: 1 dummy-path
   * invoice_blindedpay: 1 dummy-payinfo
   * invoice_created_at: now
   * invoice_payment_hash: 34567...
   * signature: SIG(^^, offer_node_id)

### request-invoice: Rusty buys 1000 sats from Matt

1. Matt generates invoice request (lnr...)

   * offer_description: 25c worth of bitcoin
   * offer_issuer: BlueMatt
   * invoice_request_payer_id: matt-nodeid

2. Rusty scans this, sends onionmessage to matt-nodeid... with reply_path back Rusty (for errors).  App shows the random invoice code for validation.

   * offer_description: 25c worth of bitcoin
   * offer_issuer: BlueMatt
   * invoice_request_payer_id: matt-nodeid
   * invoice_paths: At least 1 path
   * invoice_blindedpay: As many payinfo as invoice_path
   * invoice_created_at: now
   * invoice_payment_hash: 34567...
   * invoice_code: AV2PEE
 
3. Matt doesn't bother checking the invoice code and just hits "OK
   pay", because Rusty's right next to him and he assumes nobody
   intercepted the QR code.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -45,18 +45,19 @@ selling or has sold something.
+1. The merchant publishes an *invoice_request* which contains offer fields
+   which refer to its attempt to send money, including an amount.
+2. The user sends an *invoice* over the lightning network for the amount in the
+   *invoice_request*, using a temporary *invoice key*.
```
nit: invoice code*

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -391,65 +407,88 @@ for [Signature Calculation](#signature-calculation).
-  - MUST remember the secret key corresponding to `invoice_request_payer_key`.
-  - MUST NOT set or imply any `invoice_request_chain` not set or implied by `offer_chains`.
-  - MUST set `signature`.`sig` as detailed in [Signature Calculation](#signature-calculation) using the `invoice_request_payer_key`.
   - if `offer_quantity_min` or `offer_quantity_max` are present:
     - MUST set `invoice_request_quantity`
     - MUST set it within that (inclusive) range.
   - otherwise:
     - MUST NOT set `invoice_request_quantity`
```
I think these belong under `   - if it is responding to an offer:`

rustyrussell replied:
Well, it's still OK, it's just this never applies in that case.  But i'll move for clarify!

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -463,6 +502,12 @@ the invoice request amount exceeds the amount it's expecting (i.e. its
+
 # Invoices
 
 Invoices are a payment request, and when the payment is made, 
```
Outdated `offer_send_invoice` reference below and in the invoice's `types` section

valentinewallace reviewed (commented):
Looks reasonable to me! `send_invoice` did previously feel a little shoved in, it feels more fleshed out this way? Would need Jeff's take for which is more encoding-friendly.

valentinewallace commented:
After giving this another glance, my one doubt is that it seems like `invoice_request`s contain strictly more fields and thus may be harder to QR scan than the original `Offer` with simply `send_invoice` set

jkczyz commented:
> After giving this another glance, my one doubt is that it seems like `invoice_request`s contain strictly more fields and thus may be harder to QR scan than the original `Offer` with simply `send_invoice` set, IIUC

IIUC, the QR code would be for the `invoice_request` in the second example, which would be comparable in size to an `offer`. The first example is the typical flow where only the `offer` is communicated via QR code.

valentinewallace commented:
> > After giving this another glance, my one doubt is that it seems like `invoice_request`s contain strictly more fields and thus may be harder to QR scan than the original `Offer` with simply `send_invoice` set, IIUC
> 
> IIUC, the QR code would be for the `invoice_request` in the second example, which would be comparable in size to an `offer`. The first example is the typical flow where only the `offer` is communicated via QR code.

Hmm ok, looking in more detail it seems like it depends how you implement it. 
It looks like `offer`s have `offer_currency` and an extra `*_quantity_*` field that `invoice_request`s don't have, and `invoice_request`s have `invoice_request_payer_info` and `invoice_request_payer_note` that `offer`s don't have. 

So you could leave `offer_metadata` and `invoice_request_payer_note` blank, and they would be ~the same size with `invoice_request` being a bit smaller. But if we'll need to set `offer_metadata`, then `offer`s are nontrivially smaller. Maybe something to discuss further. 

Sorry I can't paste this with formatting but here's how I drew out the fields: 

<img width="374" alt="Screen Shot 2022-08-24 at 12 38 34 PM" src="https://user-images.githubusercontent.com/10647624/186474490-fbaaa609-5179-4d88-992e-e5ea5fc0850c.png">

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -309,20 +309,36 @@ validate the other fields.
+the onion message.
+
+The second case is publishing an `invoice_request` without an offer,
+such as via QR code.  The contains no `offer_node_id` (using the
```
s/The/It

jkczyz commented:
> So you could leave `offer_metadata` and `invoice_request_payer_note` blank, and they would be ~the same size with `invoice_request` being a bit smaller. But if we'll need to set `offer_metadata`, then `offer`s are nontrivially smaller. Maybe something to discuss further.

That's a good point. May be useful to calculate how large a typical `invoice_request` would be in this use case compared to what maximum QR code size a typical camera could support. Presumably, some fields like `invoice_request_chain` typically would be inferred rather than explicitly set. But fields like `offer_metadata` may well be needed for another 55 bytes assuming a 32-byte HMAC (1-byte type, 1-byte length, 32 byte value, times 1.6 for bech32).

valentinewallace commented on 12-offer-encoding.md:
```diff
@@ -391,65 +407,88 @@ for [Signature Calculation](#signature-calculation).
 
 The reader:
-  - MUST fail the request if `invoice_request_payer_key` is not present.
+  - MUST fail the request if `invoice_request_payer_id` is not present.
```
@rustyrussell This shouldn't hold if the invoice is in response to an `inv_req` QR, right? I believe the `invoice_request_payer_id` is optional in that case since there's no signature (assuming blinded paths are present)? 

@jkczyz if this is correct, then I think the resolution of https://github.com/rustyrussell/lightning-rfc/pull/11#issuecomment-1226526354 is that we'll never set these fields for QRs and always just set `offer_metadata`, so `invoice_request` ends up being a bit smaller than `offer_send_invoice`s but comparable

rustyrussell replied:
I ended up dropping the whole "node_id is optional if you use a blinded path and don't have a real node id" since it just made this complicated.  So that would apply here too, and it might be important later if you want to later prove your id?  It's certainly harmless to include one.

valentinewallace replied:
Hmmm, aren't we trying to save bits on the QR code here? If I'm missing something, though, I think the requirements for the `writer` need updating for this in the QR case

rustyrussell replied:
Once you have blinded paths in the QR code, it's hard to argue that saving 35 bytes is a big win.  It's possible to allow this case later, but for now we have more than enough complexity :(

valentinewallace replied:
35 bytes is pretty significant in a QR code, plus even a single 2-hop blinded path can add a lot of privacy so I'm not convinced that "blinded paths == we should give up entirely on the QR case" IMHO.

Not gonna die on this hill but wanted to put out those points :p it's true we *can* add it later as you say

jkczyz replied:
FWIW, it's actually 56 bytes after bech32 encoding, IIUC.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -391,65 +407,88 @@ for [Signature Calculation](#signature-calculation).
+        - MUST fail the request if `invoice_request_amount`.`msat` is less than the *base invoice amount*.
+        - MAY fail the request if `invoice_request_amount`.`msat` exceeds the *base invoice amount*.
+        - MUST use `invoice_request_amount`.`msat` as the *base invoice amount*.
+    - otherwise (no `offer_amount`):
+      - MUST fail the request if it does not contain `invoice_request_amount`.
```
This conflicts with the writer section:
```
    - if `offer_amount` is present:
      - MUST specify `invoice_request_amount`.
    - otherwise:
      - MAY omit `invoice_request_amount`.
```

rustyrussell replied:
Good catch!  Writer section is backwards!  "if `offer_amount` is NOT present", i.e. if offer didn't indicate an amount, invoice request MUST.

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -45,18 +45,19 @@ selling or has sold something.
-2. The user sends an *invoice* for the amount in the *offer*
-3. The merchant makes a payment to the user indicated by the invoice.
+1. The merchant publishes an *invoice_request* which contains offer fields
+   which refer to its attempt to send money, including an amount.
```
nit: s/which/that

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -391,65 +407,88 @@ for [Signature Calculation](#signature-calculation).
+      - MUST specify `invoice_request_amount`.
+    - otherwise:
+      - MAY omit `invoice_request_amount`.
+      - if it sets `invoice_request_amount`:
+        - MUST specify `invoice_request_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency`).
```
Shouldn't this only apply if `offer_amount` is present? (i.e, not under "otherwise:")

rustyrussell replied:
Yes, this was backwards!

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -391,65 +407,88 @@ for [Signature Calculation](#signature-calculation).
+          `invoice_request_chain` currency.
+        - if `invoice_request_quantity` is present, multiply by `invoice_request_quantity`.`quantity`.
+      - if `invoice_request_amount` is present:
+        - MUST fail the request if `invoice_request_amount`.`msat` is less than the *base invoice amount*.
+        - MAY fail the request if `invoice_request_amount`.`msat` exceeds the *base invoice amount*.
+        - MUST use `invoice_request_amount`.`msat` as the *base invoice amount*.
```
In the writer section, it isn't clear that the quantity must be accounted for when setting `invoice_request_amount`. In fact, it even says:

> - MUST specify `invoice_request_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency`).

Which seems to imply the quantity _shouldn't_ be accounted for when setting `invoice_request_amount`. But here it implies that it is. Am I missing something?

(IMHO, it would be easier to read and less error prone if valid values were described in one place instead of across reader and writer sections.)

rustyrussell replied:
It might be harder to write this way, but it's far less error prone to implement.  values don't have requirements, readers and writers do.

The non-Requirements sections are useful to give context, but the requirements must be written this way, as that is what implementers need to follow.

jkczyz replied:
Looks like only the parenthetical was addressed. Could you take a look at my original question?

jkczyz commented on 12-offer-encoding.md:
```diff
@@ -605,30 +655,21 @@ A writer of an invoice:
-      - MAY simply reuse the previous invoice.
-    - otherwise:
-      - MUST NOT reuse a previous invoice.
-    - if `invoice_request_amount` is present:
-      - MUST set `invoice_amount` to `invoice_request_amount`
-    - otherwise: (no `invoice_request_amount`, so must have `offer_amount`):
-      - MUST set `invoice_amount`.`msat` to the *base invoice amount*.
```
Why drop this?

rustyrussell replied:
Previously, the base invoice amount was calculated in the invoice_request parsing, and *used* in invoice writing.  Which was just weird.

Now we reflect all the fields, the invoice writing has all the information to calculate this where it's needed.  So it was moved.

rustyrussell committed:
BOLT 12: remove `send_invoice` offer, use a straight `invoice_request`.

Instead of a kind of dummy offer which asks you to send an invoice,
use a dummy invoice request, since it now contains all the offer
fields anyway.  It simplifies things a little, but there's the
following fallout:

1. We use the presence of `offer_node_id` to distinguish
   invoice-request-for-offer and invoice-request-no-offer, which makes
   sense, since the vendor doesn't know that.
2. We rename `invoice_request_payer_key` to `invoice_request_payer_id`.  In
   the case of a public node making a non-offer invoice_request, it's
   their id (as they are the payer), but it can omitted using blinded
   paths (like the offer_node_id in offer).
3. The invoice_request has no signature if it's not an offer response.
   This is makes them as small like the old send_invoice offer.
4. We don't allow invoice_request in other currencies (could in future):
   must specify in msats (this issue is expanded in the error section).
5. Actually tell them they're supposed to request an invoice when they
   receive an `invoice_request` for an offer!  (They don't have to if it's
   a non-offer one, they would be expected to check with user).
6. No signatures on invoices if they're non-offer, but allow a verification
   code.  If the user presents an offer to the ATM in the first place, this
   can be skipped, but requires it have a camera, and more interaction.

Here are examples of the offer-request-invoice flow, and the
request-invoice flow:

### offer-request-invoice: Rusty buys a coffee from Starblocks

1. Scans lno...

   * offer_description: 1 cup coffee.
   * offer_amount: 1000000msat
   * offer_issuer: Starblocks starblocks.acinq.co
   * offer_node_id: starblocks-nodeid

2. Rusty sends onionmessage to starblocks-nodeid with reply_path back Rusty, and
invoice_request:

   * invoice_request_metadata: 23456...
   * offer_description: 1 cup coffee.
   * offer_amount: 1000000msat
   * offer_issuer: Starblocks starblocks.acinq.co
   * offer_node_id: starblocks-nodeid
   * invoice_request_payer_id: rusty-tmpkey
   * invoice_request_payer_note: love that coffee!
   * signature: SIG(^^, rusty-tmpkey)
   
3. Starblocks replies with invoice:

   * invoice_request_metadata: 23456...
   * offer_description: 1 cup coffee.
   * offer_amount: 1000000msat
   * offer_issuer: Starblocks starblocks.acinq.co
   * offer_node_id: starblocks-nodeid
   * invoice_request_payer_id: rusty-tmpkey
   * invoice_request_payer_note: love that coffee!
   * invoice_paths: 1 dummy-path
   * invoice_blindedpay: 1 dummy-payinfo
   * invoice_created_at: now
   * invoice_payment_hash: 34567...
   * signature: SIG(^^, offer_node_id)

### request-invoice: Rusty buys 1000 sats from Matt

1. Matt generates invoice request (lnr...)

   * offer_description: 25c worth of bitcoin
   * offer_issuer: BlueMatt
   * invoice_request_payer_id: matt-nodeid

2. Rusty scans this, sends onionmessage to matt-nodeid... with reply_path back Rusty (for errors).  App shows the random invoice code for validation.

   * offer_description: 25c worth of bitcoin
   * offer_issuer: BlueMatt
   * invoice_request_payer_id: matt-nodeid
   * invoice_paths: At least 1 path
   * invoice_blindedpay: As many payinfo as invoice_path
   * invoice_created_at: now
   * invoice_payment_hash: 34567...
   * invoice_code: AV2PEE
 
3. Matt doesn't bother checking the invoice code and just hits "OK
   pay", because Rusty's right next to him and he assumes nobody
   intercepted the QR code.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
Feedback from @valentinewallace

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
BOLT 12: more feedback, and remove stray tabs.

Most confusingly, the "if `offer_amount` is present" case was
exactly backwards, which doesn't make sense at all.

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell commented:
Merged into bolt12 PR.  Thanks!
